### PR TITLE
improvement: show return type annotations in fixtures

### LINKED ISSUE #13676: Show return type annotations in `--fixtures` and `--fixtures-per-test`
#### What's the problem this feature will solve?
Knowing the annotated return types of fixtures in `--fixtures` could help users to understand what a fixture is/does, especially for fixtures that are missing a docstring.

#### Describe the solution you'd like

Something like

```python
import pytest

@pytest.fixture
def answer() -> int:
    """Answer to the Ultimate Question of Life, the Universe, and Everything."""
    return 42
```

currently displays:

```
------------------------- fixtures defined from test_x -------------------------
answer -- test_x.py:4
    Answer to the Ultimate Question of Life, the Universe, and Everything.
```

but could display:

```
------------------------- fixtures defined from test_x -------------------------
answer -> int -- test_x.py:4
    Answer to the Ultimate Question of Life, the Universe, and Everything.
```

or

```
------------------------- fixtures defined from test_x -------------------------
answer -- test_x.py:4
    Answer to the Ultimate Question of Life, the Universe, and Everything.
    Return type: int
```

or any other [color of bike shed](https://shed.bike/).

#### Alternative Solutions
Looking at the source code at the indicated location. Works, but is an extra step.

#### Additional context
I'm a bit confused between [`typing.get_type_hints()`](https://docs.python.org/3.14/library/typing.html#typing.get_type_hints), [`inspect.get_annotations()`](https://docs.python.org/3.14/library/inspect.html#inspect.get_annotations) (pre-3.14) and [annotationlib](https://docs.python.org/3.14/library/annotationlib.html#annotationlib.get_annotations) (3.14+), and which one would be best to use for this purpose.

### PR DESCRIPTION
**Pull request by kianelbo**
closes #13676

### CODE REVIEW

File: src/_pytest/fixtures.py [hunk 1]
```diff
@@ -2013,6 +2019,17 @@ def _showfixtures_main(config: Config, session: Session) -> None:
         tw.line()
 
 
+def get_return_annotation(fixture_func: Callable[..., Any]) -> str:
+    try:
+        sig = signature(fixture_func)
+        annotation = sig.return_annotation
+        if annotation is not sig.empty and annotation != inspect._empty:
+            return inspect.formatannotation(annotation).replace("'", "")
```
Comment (The-Compiler) [inline src/_pytest/fixtures.py]: This seems odd to me, for various reasons:

- `inspect._empty` is private so it shouldn't be used
- `inspect.formatannotation`, while public, is undocumented so probably shouldn't be used either
- Why remove `'` from it, which is something `inspect` itself doesn't seem to do?

File: changelog/13676.improvement.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Added return type annotations in ``fixtures`` and ``fixtures-per-test``.
```
Comment (The-Compiler) [inline changelog/13676.improvement.rst]:
```suggestion
Added return type annotations in ``--fixtures`` and ``--fixtures-per-test``.
```

**Review by The-Compiler — CHANGES_REQUESTED** [review]

File: src/_pytest/fixtures.py (hunk 1 above)
Comment (kianelbo) [reply src/_pytest/fixtures.py]: Yes, that `inspect._empty` condition is actually redundant. I investigated the logic further and there were lots of inconsistent behaviors in the test (that's why I previously did `replace("'", "")`) and corner cases. However, the new solution (though a bit hacky) is the most general one and hopefully covering everything.

File: src/_pytest/fixtures.py [hunk 3]
```diff
@@ -2013,6 +2019,21 @@ def _showfixtures_main(config: Config, session: Session) -> None:
...
+    try:
+        sig = signature(fixture_func)
+        annotation = sig.return_annotation
+        if annotation is not sig.empty:
+            if isinstance(annotation, str):
+                return annotation
+            if annotation.__module__ == "typing":
+                return str(annotation).replace("typing.", "")
+            return annotation.__name__
+    except (ValueError, TypeError):
```
Comment (The-Compiler) [inline src/_pytest/fixtures.py]: Why would `ValueError` and `TypeError` happen here? The only thing that could happen I think is `AttributeError`, which already happens with e.g. `-> None:` as annotation.

That being said, `None` should definitively handled correctly by the code and tested properly as well.

File: testing/python/fixtures.py [hunk 4]
```diff
@@ -5068,3 +5088,45 @@ def test_method(self, /, fix):
     )
     result = pytester.runpytest()
     result.assert_outcomes(passed=1)
+
+
+def test_get_return_annotation() -> None:
```
Comment (The-Compiler) [inline testing/python/fixtures.py]: Maybe turn this into a test class and split the individual asserts into separate test functions? That way, if one of them fails, the rest of the tests will still run.

**Review by The-Compiler — CHANGES_REQUESTED** [review]

File: testing/python/fixtures.py [hunk 5]
```diff
@@ -5068,3 +5088,45 @@ def test_method(self, /, fix):
...
+        return two_sixes
+
+    assert get_return_annotation(callable_return) == "Callable[..., Any]"
+
+    def no_annotation():
+        return 6
+
+    assert get_return_annotation(no_annotation) == ""
+
+    def none_return() -> None:
```
Comment (The-Compiler) [inline testing/python/fixtures.py]: Ah, this test happens to work (despite the bug above) because this module uses `from __future__ import annotations`, so the annotation already is `'None'` rather than `None`.

Not sure how to best get around this. Maybe those tests should be in a separate module which deliberately doesn't use that?

File: src/_pytest/fixtures.py (hunk 3 above)
Comment (kianelbo) [reply src/_pytest/fixtures.py]: TypeError happens when the argument is not a callable (e.g. a number). I doubt this case happens. ValueError happens when the passed argument is a callable but a signature can't be obtained (e.g. `range`). Also unlikely imo.
For `-> None:` on my local test didn't raise any error.

File: testing/python/fixtures.py (hunk 5 above)
Comment (kianelbo) [reply testing/python/fixtures.py]: On my local test (3.13) with or without `from __future__ import annotations` it worked the same. I might be very wrong but I think it behaves differently in different Python versions. Still, I added a exclusive check for none. Do you think moving it to another module is worth it? Seems too much for a humble helper function!

PS: The reason I didn't use `isinstance(annotation, types.NoneType) was that pylance was complaining.

**Comment by kianelbo** [comment]
@The-Compiler Could you please review again?
And any tips to satisfy codecov? The indicated uncovered lines are mostly test codes themselves.

**Comment by The-Compiler** [comment]
Sorry for the radio silence, I never got around to taking a deeper look at this, seems like getting a type annotation as a string is surprisingly difficult with a lot of corner cases...

I still think having a file without the `__future__` import is absolutely crucial for those tests - because as-is, they don't actually test most of your code, and [that shows in coverage](https://app.codecov.io/gh/pytest-dev/pytest/pull/13680).

Trying with `-> tuple[int, str]` in a new file, all that's displayed is `tuple`, and I'm pretty sure after looking at the implementation of `inspect.formatannotation` that there are more funny corner cases that possibly cause a crash currently (e.g. `-> 42` currently does - while not a valid annotation semantically, it is valid Python).

Given the amount of corner cases, I wonder if (despite what I said earlier) we should go back at letting `inspectlib` handle those. Perhaps we can lessen the future impact of using an undocumented function by passing `annotation_format=annotationlib.Format.STRING` to `inspect.signature` with Python >= 3.14, then we can take an easy shortcut there and simply return the string directly in that case.

#13550 added a helper `signature` method that does this already, so I'd recommend basing things on that.

File: src/_pytest/fixtures.py (hunk 3 above)
Comment (The-Compiler) [reply src/_pytest/fixtures.py]: Both of those cause neither of those exceptions, but cause an `AttributeError` for `.__module__`.

**Comment by kianelbo** [comment]
> Given the amount of corner cases, I wonder if (despite what I said earlier) we should go back at letting `inspectlib` handle those. Perhaps we can lessen the future impact of using an undocumented function by passing `annotation_format=annotationlib.Format.STRING` to `inspect.signature` with Python >= 3.14, then we can take an easy shortcut there and simply return the string directly in that case.

Are you suggesting to use `formatannotation` again?

**Comment by The-Compiler** [comment]
At least for Python < 3.14, for two reasons:

- A implementation of this considering all possible corner cases evidently [is trickier than I first thought](https://github.com/python/cpython/blob/3.13/Lib/inspect.py#L1439-L1451) (and just copy-pasting that code into pytest isn't really better I think)
- If we can use a more proper way on Python 3.14 (simply getting the annotations as strings in the first place, and using them directly), the risk of Python removing the undocumented function in the future is minimal, as I doubt such a thing would happen in a patch release of Python.

**Comment by kianelbo** [comment]
In addition to the current code, I've tried `inspect.get_annotation` (with or without `formatannotation`) and `typing.get_type_hints` and even with lots of tinkering they all fail miserably on the test cases. It seems impossible! There's another approach, parsing with `ast`, which works decently, however, the tests fail because it can't access the source code. Any suggestions @The-Compiler ?

**Comment by The-Compiler** [comment]
I suspected this is going to be trickier than it looked on the surface :sweat_smile:. I really hope we don't need to resort to AST shenanigans, after all, other tools do sometimes print type annotations as well I think?

Can you elaborate on "all fail miserably on the test cases"? I'd hope that `inspect.signature` (or `inspect.get_annotations` which is internally used by `inspect.signature`) and `formatannotation` (+ forcing a string on Python 3.14+) should do the right thing here, but it sounds like that's not the case?

**Comment by kianelbo** [comment]
> Can you elaborate on "all fail miserably on the test cases"?

I mean the test cases fail. For example doing:
```
try:
        annotation = inspect.get_annotations(fixture_func)["return"]
        return inspect.formatannotation(annotation)
    except (KeyError, ValueError, TypeError):
        pass
    return ""
```
will give `collections.abc.Callable[..., typing.Any]` instead of `Callable[..., Any]` in the test case. The more surprising thing is that when I run the code in IPython for testing, it produces the desired output `Callable[..., Any]`.

**Comment by The-Compiler** [comment]
I think that's fine, after all that's what `inspect.signature` does as well:

```python
import inspect
from collections.abc import Callable

def func(x: Callable[[], None]) -> None:
    pass

print(inspect.signature(func))
```

And with the approach outlined above, starting on Python 3.14 (or before with `from __future__ import annotations`) it will print whatever is there in the source file anyways, so I wouldn't worry about it too much.

**Comment by kianelbo** [comment]
The best that I could come up with was treating signature as string (as you suggested) and use regex to remove clutterings (if present) and it passed the tests locally. But again the tests are failing in CI and codecov is complaining unfairly. Any tips?

**Comment by The-Compiler** [comment]
That's not what I meant with treating it as a string at all :sweat_smile:

Parsing the string output definitely isn't necessary here, as you're just picking out what was rendered using `formatannotation`. Despite the caveat of that being undocumented, calling it is definitely a lot better than parsing things from the string.

What I meant was that with Python 3.14, you can pass `annotation_format=annotationlib.Format.STRING` to `inspect.signature` (or probably better to use `annotationlib.get_annotations()` at this point, as you need to import it anyways and we don't need anything from `inspect`). Then from my understanding, you always get a string back, and you don't need to format it any way.

So basically I see three scenarios the code needs to handle:

- Python < 3.14, no future import (annotations are objects): Use `inspect.signature` (or `inspect.get_annotations()`, doesn't matter too much) and the (undocumented but stable hopefully) `formatannotation`.
- Python < 3.14, string annotations (literally or via `from __future__ import annotations`): You might want to skip `formatannotation` and just use it directly.
- Python >= 3.14 (annotations are always strings): Use `annotationlib.get_annotations()` with `annotationlib.Format.STRING` and use the resulting string directly.

As for the coverage, looks like pre-commit tried to be helpful and introduced the `__future__` import again: https://github.com/pytest-dev/pytest/pull/13680/commits/1c18e591c60923ba25ec5312f6fa1f038c49633e

This seems to happen via ruff: 

https://github.com/pytest-dev/pytest/blob/a55c959e4d5e5770ed5b3433d9d3714d63b700f3/pyproject.toml#L102

So if you add a `# ruff: noqa: FA100` to the top of that file, that should hopefully stop happening ([docs](https://docs.astral.sh/ruff/linter/#error-suppression)).

**Comment by kianelbo** [comment]
Understood, but `formatannotation` doesn't work well, for example it returns:
`test_custom_class_return_type.<locals>.T` instead of just `T`. That's why I used the regex.

**Comment by The-Compiler** [comment]
That's because it formats the `__module__` and the [`__qualname__`](https://docs.python.org/3/glossary.html#term-qualified-name) into the output.

Maybe that could actually be useful information in real-world scenarios (try e.g. with `class T` being defined at module level instead of inside a function), but yeah, it could also be too verbose.

It looks like `formatannotation` also [supports](https://github.com/python/cpython/blob/3.13/Lib/inspect.py#L1439-L1451) a `base_module` you can pass to avoid including that part if the class is defined in the same module the fixture is (by passing the fixture module as base).

If you disagree, I suppose we're back to copying the [`formatannotation` code](https://github.com/python/cpython/blob/3.13/Lib/inspect.py#L1439-L1451), and adjusting it so that it uses `annotation.__name__` in case the annotation is a `type` (i.e. a class). Would be fine in my book as well, as it's only some 10 lines.

**Comment by kianelbo** [comment]
> It looks like `formatannotation` also [supports](https://github.com/python/cpython/blob/3.13/Lib/inspect.py#L1439-L1451) a `base_module` you can pass to avoid including that part if the class is defined in the same module the fixture is (by passing the fixture module as base).

I've tried it but it doesn't work as intended. Do you object using regex to remove the extra verbose parts?

> So if you add a # ruff: noqa: FA100 to the top of that file, that should hopefully stop happening ([docs](https://docs.astral.sh/ruff/linter/#error-suppression)).
[Didn't work!](https://github.com/pytest-dev/pytest/pull/13680/commits/f555682bb33f37729782a26dd35d0e7fc7841342)

**Comment by webknjaz** [comment]
@kianelbo this has gotten stale. Would you mind rebasing?

> [!tip]
> *Pro tip*: in the future, it's best to file PRs from feature branches.
>
> https://hynek.me/articles/pull-requests-branch/

File: src/_pytest/fixtures.py [hunk 6]
```diff
@@ -2009,6 +2016,18 @@ def _showfixtures_main(config: Config, session: Session) -> None:
         tw.line()
 
 
+def get_return_annotation(fixture_func: Callable[..., Any]) -> str:
+    pattern = re.compile(r"\b(?:(?:<[^>]+>|[A-Za-z_]\w*)\.)+([A-Za-z_]\w*)\b")
+
+    sig = signature(fixture_func)
```
Comment (RonnyPfannschmidt) [inline src/_pytest/fixtures.py:2022]: We need to avoid resolving unresolvable type annotations there 

This is an issue with python 3.14 on certain codebases

File: src/_pytest/fixtures.py [hunk 7]
```diff
@@ -2009,6 +2016,18 @@ def _showfixtures_main(config: Config, session: Session) -> None:
         tw.line()
 
 
+def get_return_annotation(fixture_func: Callable[..., Any]) -> str:
+    pattern = re.compile(r"\b(?:(?:<[^>]+>|[A-Za-z_]\w*)\.)+([A-Za-z_]\w*)\b")
```
Comment (RonnyPfannschmidt) [inline src/_pytest/fixtures.py:2020]: This regex needs a explanation and if we compile it we should do that either once or cached

**Review by RonnyPfannschmidt — CHANGES_REQUESTED** [review]
Thanks for getting this started 

This looks readable 

Why was a new file created and was an ai assistant involved

**Comment by RonnyPfannschmidt** [comment]
OK I'm done with github mobile its ui glitches get me every time

File: src/_pytest/fixtures.py (hunk 6 above)
Comment (nicoddemus) [reply src/_pytest/fixtures.py:2022]: Do you mean sometimes `signature` will raise an error, depending on the case? Could you provide an example, so we can add a test for it?

**Comment by nicoddemus** [comment]
> Understood, but formatannotation doesn't work well, for example it returns:
test_custom_class_return_type.<locals>.T instead of just T. That's why I used the regex.

Btw I'm OK with just displaying what `formatannotation` returns, without any regex matching. Might be a bit cluttered, but keeps the code simple and we don't hide possible useful information to the user.

File: src/_pytest/fixtures.py (hunk 6 above)
Comment (RonnyPfannschmidt) [reply src/_pytest/fixtures.py:2022]: Yes this also is a issue in pluggy but there is can sidestep the issue differently
